sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,133 @@
+
+#### Relevant issues/links
+
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+
+#### Possible Solution(s)
+
+- Moving from iptables to IPVS may help here if that works stably. We have the IPVS alternative [implemented](https://github.com/kubernetes/kubernetes/pull/46580), but it still hasn’t gone to GA.
+- For a long time the official kernel upstream answer for this issue was that nftables was going to solve all iptables-related scalability problems (both with packet processing and with rule changes), and there's now an out-of-tree nftables kube-proxy backend too ([#62720](https://github.com/kubernetes/kubernetes/issues/62720)).
+- There's also an alternative plan to fix the packet processing speed problems by rewriting iptables to use eBPF inside the kernel (see - https://cilium.io/blog/2018/04/17/why-is-the-kernel-community-replacing-iptables).
+
+
+## Slow/failing iptables-restore operations when large number of rules exist
+
+#### Issue
+
+When we’re running services worth a large number of backends (> 100k) on our large-cluster scalability tests, we’re noticing that kube-proxy is timing out while trying to do iptables-restore due to failing to acquire lock over iptables. There are at least two parts to this problem:
+
+- If a process tried to do a very large iptables-restore, and other processes on the system were also doing iptables-related things, the iptables-restore might get restarted multiple times (inside the kernel) before completing, which could then cause another iptables operation running in another goroutine to time out before it ever got the xtables lock.
+- The `iptables-restore` implementation is such that it grabs the lock before it parses its input, and simply parsing tens of thousands of iptables rules takes a noticeable amount of time. So if two iptables commands start at the same time, the first might grab the lock, and then start parsing its input, and burn through half of the other iptables command's `--wait` time before it even gets to the point of passing the rules off to the kernel.
+
+TODO: Find if there's any issue arising here from k8s side (relevant bug linked below).
+
+#### Relevant issues/links
+
+- [#48107](https://github.com/kubernetes/kubernetes/issues/48107)
+
+#### Possible Solution(s)
+
+For the first problem:
+
+- There's [a claim](https://github.com/kubernetes/kubernetes/issues/48107#issuecomment-398081930) that upgrading to linux kernel version 4.15 may fix this. Though we haven't verified this (current COS versions using <= 4.14 kernel version).
+
+For the second problem:
+
+- It's not easy to fix it because of the way that `iptables-restore` was written to share code with the main `iptables` binary, and no one is working on fixing it because the official plan is to move to nftables instead. If the world ends up moving to iptables-over-eBPF rather than nftables then this will probably need to be fixed at some point.
```

Comment:
In general, if our only problem are timeouts, maybe it would be enough to bump them?

I guess I'm saying that speeding this up may be difficult, but getting rid of errors sounds like something we may be able to do.

Reply:
Note that keep using iptables interface it will inherit all the interface problems that iptables has. The evolution to nftables is not a whim, it was designed to fulfill all the lacks that iptables interface has but also much more features that permits to implement more use cases with better performance.